controllers/storageclassclaim/storageclassclaim_controller.go

```diff
@@ -310,8 +319,60 @@ func (r *StorageClassClaimReconciler) reconcileConsumerPhases() (reconcile.Resul
 		// Update the StorageClassClaim status.
 		r.storageClassClaim.Status.Phase = v1alpha1.StorageClassClaimReady
 
+	} else if r.storageClassClaim.UID != "" {
+
+		// Deletion phase.
+		// Update the StorageClassClaim status.
+		r.storageClassClaim.Status.Phase = v1alpha1.StorageClassClaimDeleting
+
+		// Call `RevokeStorageClassClaim` service on the provider server with StorageClassClaim as a request message.
+		// Check if StorageClassClaim is still exists (it might have been manually removed during the StorageClass
+		// removal above).
+		_, err = providerClient.RevokeStorageClassClaim(r.ctx, r.storageCluster.Status.ExternalStorage.ConsumerID, r.storageClassClaim.Name)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+
+		// Delete StorageClass.
+		// Make sure there are no StorageClass consumers left.
+		// Check if StorageClass is in use, if yes, then fail.
+		// Wait until all PVs using the StorageClass under deletion are removed.
+		// Check for any PVs using the StorageClass.
+		pvList := corev1.PersistentVolumeList{}
+		err := r.list(&pvList, client.MatchingFields{
+			"spec.storageClassName": r.storageClassClaim.Name,
+		})
+		if err != nil {
+			return reconcile.Result{}, fmt.Errorf("failed to list PersistentVolumes: %s", err)
+		}
+		if len(pvList.Items) > 0 {
+			return reconcile.Result{}, fmt.Errorf("StorageClass %s is still in use by one or more PV(s)",
+				r.storageClassClaim.Name)
+		}
+		storageClass := &storagev1.StorageClass{}
+		storageClass.Name = r.storageClassClaim.Name
+		if err = r.get(storageClass); err == nil {
+
+			// Remove finalizer from StorageClass.
+			storageClass.SetFinalizers(remove(storageClass.GetFinalizers(), StorageClassFinalizer))
+			if err := r.update(storageClass); err != nil {
+				return reconcile.Result{}, fmt.Errorf("failed to remove finalizer from StorageClass: %s", err)
+			}
+			if err = r.delete(storageClass); err != nil {
+				return reconcile.Result{}, fmt.Errorf("failed to delete StorageClass %s: %s", storageClass.Name, err)
+			}
+		} else if errors.IsNotFound(err) {
+			r.log.Info("StorageClass already deleted.")
+		} else {
+			return reconcile.Result{}, fmt.Errorf("failed to get StorageClass %s: %s", storageClass.Name, err)
+		}
+
```

Comment:
From a readability standpoint, it will be better to handle the errors near the get operation and to move the main code after the error handling.

```suggestion
		if err = r.get(storageClass); err != nil &&  !errors.IsNotFound(err){ 
			return reconcile.Result{}, fmt.Errorf("failed to get StorageClass %s: %s", storageClass.Name, err)
		}
		if (storageClass.UID != "") {
			// Remove finalizer from StorageClass.
			storageClass.SetFinalizers(remove(storageClass.GetFinalizers(), StorageClassFinalizer))
			if err := r.update(storageClass); err != nil {
				return reconcile.Result{}, fmt.Errorf("failed to remove finalizer from StorageClass: %s", err)
			}
			if err = r.delete(storageClass); err != nil {
				return reconcile.Result{}, fmt.Errorf("failed to delete StorageClass %s: %s", storageClass.Name, err)
			}
		} else {
			r.log.Info("StorageClass already deleted.")
		}
```